### PR TITLE
feat: added ros-time with simulated time to library

### LINKED ISSUE #7: Implement ros Clock

### PR DESCRIPTION
**Pull request by tdejager**
Closes #7 

This implements the simulated ROS-time for Rosty, with corresponding tests. There is a `now` function in the public API that either gives the current time since epoch or the simulated clock time depending on if the parameter `/use_sim_time` is set. Note, that this API is a bit of a departure from classical Ros libraries, that implement this method on Time.

My reasons are because it is actually a function with side-effects and that can potentially panic, that it is nicer to place this directly in the public API. Because, putting it in the Time struct, or in our case, the Duration struct, would imply that it would only use the System time. Which is less likely to fail.

I've also made all the test utility functions public, otherwise, warnings would be issued during compilation.

### CODE REVIEW

File: crates/rosty/src/node.rs [hunk 1]
```diff
@@ -131,6 +134,20 @@ impl Node {
             *mutex_guard = Some(tokio::try_join!(slave_future).map(|_| ()))
         });
 
+        // Check if we need to use simtime
+        let param = Param::new("/use_sim_time", master.clone());
+
+        // Try to get the sim_time, and open a topic if we are waiting for it
+        let sim_time = if param.exists().await? {
```
Comment (baszalmstra) [inline crates/rosty/src/node.rs:141]: Why not use && here?

File: crates/rosty/src/node.rs [hunk 2]
```diff
@@ -139,9 +156,21 @@ impl Node {
...
             result: result_mutex,
             shutdown_token,
+            sim_time,
         })
     }
 
+    /// Initializes the receiving of the simulated time
+    /// This function panics if the '/use_sim_time'
+    /// parameter was not found
+    pub async fn init_sim_time(&self) -> Result<(), SubscriptionError> {
```
Comment (baszalmstra) [inline crates/rosty/src/node.rs:166]: Maybe make this pub(crate)

File: crates/rosty/src/node.rs [hunk 3]
```diff
@@ -172,6 +201,17 @@ impl Node {
...
     }
 
+    /// Returns true if this node is using the simulated time
+    pub fn is_using_sim_time(&self) -> bool {
+        self.sim_time.is_some()
+    }
+
+    /// Returns the last simulated clock if it is available
+    pub fn get_last_sim_clock(&self) -> Option<Duration> {
+        let duration = self.sim_time.as_ref()?.duration();
```
Comment (baszalmstra) [inline crates/rosty/src/node.rs:211]: Remove let duration? is already clear from the call to duration

File: crates/rosty/src/node.rs (hunk 2 above)
Comment (baszalmstra) [reply crates/rosty/src/node.rs:166]: Ah nvm, Node itself is not exposed

File: crates/rosty/src/lib.rs [hunk 4]
```diff
@@ -40,6 +42,11 @@ pub async fn init_with_args(args: NodeArgs, capture_sigint: bool) -> Result<(),
 
     let node = Node::new(args).await?;
 
+    // Initialize the use of simtime
+    if node.is_using_sim_time() {
```
Comment (baszalmstra) [inline crates/rosty/src/lib.rs:46]: Why not do this in Node::new?

File: crates/rosty/src/simtime.rs [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+use crate::Duration;
+use futures::StreamExt;
+use rosty_msg::rosgraph_msgs::Clock;
+use std::sync::{Arc, Mutex};
+
+#[derive(Debug)]
+/// The Simulated Time struct, holds a reference to the last message received on the clock interface
+pub struct SimTime {
+    last_clock_msg: Arc<Mutex<Option<Clock>>>,
```
Comment (baszalmstra) [inline crates/rosty/src/simtime.rs:11]: Maybe use RwLock instead of Mutex?

File: crates/rosty/tests/parameters.rs [hunk 6]
```diff
@@ -28,6 +28,12 @@ fn test_parameter_api() {
         // And it should be changed
         assert_eq!(param.get::<i32>().await.unwrap(), 10);
 
+        // Change it to a boolean
```
Comment (baszalmstra) [inline crates/rosty/tests/parameters.rs:31]: Why is this removed?

File: crates/rosty/tests/shutdown_token.rs [hunk 7]
```diff
@@ -1,4 +1,4 @@
-mod util;
+pub mod util;
```
Comment (baszalmstra) [inline crates/rosty/tests/shutdown_token.rs:1]: Why does this need to be pub?

Comment (tdejager) [reply crates/rosty/tests/shutdown_token.rs:1]: So that there are no warnings when compiling. Because Rust treats every test file as a different compilation unit, you can get warnings if the functions are not being used in *each* test file.

File: crates/rosty/tests/parameters.rs (hunk 6 above)
Comment (tdejager) [reply crates/rosty/tests/parameters.rs:31]: It was broken, this needs to be fixed :)

File: crates/rosty/src/lib.rs (hunk 4 above)
Comment (tdejager) [reply crates/rosty/src/lib.rs:46]: I don't know, because it is so stateful if feels weird putting it there?

File: crates/rosty/src/simtime.rs (hunk 5 above)
Comment (tdejager) [reply crates/rosty/src/simtime.rs]: Done

File: crates/rosty/src/node.rs (hunk 3 above)
Comment (tdejager) [reply crates/rosty/src/node.rs]: Done

File: crates/rosty/src/node.rs (hunk 1 above)
Comment (tdejager) [reply crates/rosty/src/node.rs]: Done
